clarify macro properties aren't supported in dbt_project.yml

mirnawong1 opened the pull request:
After convo with jerco and [per slack convo](https://getdbt.slack.com/archives/C2JRRQDTL/p1693287414438449), clarify in the docs that macro properties aren't support in dbt_project.yml file. 

adding a note in the following pages:
 
- https://docs.getdbt.com/reference/dbt_project.yml
- https://docs.getdbt.com/reference/resource-configs/docs
- https://docs.getdbt.com/reference/macro-properties

note - not sure how semantic models/metrics are added here

Commit:
clarify macro aren't support in project.yml file

mirnawong1 committed:
Merge branch 'current' into mwong-macro-yml

jtcohen6 commented on website/docs/reference/dbt_project.yml.md:
```diff
@@ -18,12 +18,13 @@ By default, dbt will look for `dbt_project.yml` in your current working director
 dbt uses YAML in a few different places. If you're new to YAML, it would be worth taking the time to learn how arrays, dictionaries and strings are represented.
 :::
 
+The following is a list of all available configurations in the `dbt_project.yml` file. Something to note, [Macro properties](/reference/macro-properties) aren't supported in the `dbt_project.yml` file:
```
Recommend updating this note to mention that these configurations aren't supported in `dbt_project.yml`:
- Macro properties: https://docs.getdbt.com/reference/macro-properties
- Analysis properties: https://docs.getdbt.com/reference/analysis-properties
- Semantic model configs (https://docs.getdbt.com/docs/build/semantic-models) — **however** we are adding support for this in v1.7 ! (https://github.com/dbt-labs/dbt-core/pull/8502)

mirnawong1 replied:
thanks so much @jtcohen6 , this is so clear to me now so i really appreciate you breaking it down 🙏 

this issue https://github.com/dbt-labs/docs.getdbt.com/issues/3997 makes it seems that semantic models are configs, not properties -- is that right?

mirnawong1 replied:
oh wait, sorry! i see the issue is related to groups only!

jtcohen6 reviewed (commented):
Thank you for following up here @mirnawong1!

There's a distinction between attributes we call "configs" and attributes we call "properties," in that "configs" can be set for many resources at once in `dbt_project.yml`, and "properties" can only be set for one resource at a time. The fact is that macros and analyses _only_ have properties, and that's why they cannot be configured from `dbt_project.yml`.

https://docs.getdbt.com/reference/configs-and-properties

I am very supportive of the idea of calling this out explicitly in the various resource pages! I will also just say, this distinction is historical & a bit arbitrary. We have been thinking on & off about how we might change it in the future:
- https://github.com/dbt-labs/dbt-core/issues/7157

matthewshaver commented on website/docs/reference/macro-properties.md:
```diff
@@ -2,9 +2,9 @@
 ---
 
-Macro properties can be declared in `.yml` files.
+Macro properties can be declared in `.yml` files, except in the `dbt_project.yml` files.
```
```suggestion
Macro properties can be declared in any `.yml` file except for the `dbt_project.yml` file.
```

matthewshaver commented on website/docs/reference/macro-properties.md:
```diff
@@ -2,9 +2,9 @@
+Macro properties can be declared in `.yml` files, except in the `dbt_project.yml` files.
 
-You can name these files `whatever_you_want.yml`, and nest them arbitrarily deeply in subfolders.
+You can name these files `whatever_you_want.yml`, and nest them arbitrarily deeply in sub-folders.
```
```suggestion
You can name these files `whatever_you_want.yml` and nest them arbitrarily deep in sub-folders.
```

Commit:
add snippet and rephrase

mirnawong1 committed:
Update dbt_project.yml.md

mirnawong1 committed:
Merge branch 'current' into mwong-macro-yml

mirnawong1 committed:
Update website/docs/reference/macro-properties.md

Co-authored-by: Matt Shaver <60105315+matthewshaver@users.noreply.github.com>

mirnawong1 committed:
Update website/docs/reference/macro-properties.md

Co-authored-by: Matt Shaver <60105315+matthewshaver@users.noreply.github.com>

Commit:
Merge branch 'mwong-macro-yml' of https://github.com/dbt-labs/docs.getdbt.com into mwong-macro-yml

Commit:
update lang

mirnawong1 commented:
hey @dbeatty10 , can i request a tech review for this pr? so I've:

- created a partial to communicate that certain resource properties can't be set in the dbt_project.yml file. this partial is set for macro, analysis, exposure, and source properties [per this doc](https://docs.getdbt.com/reference/configs-and-properties#where-can-i-define-properties).
- added explicit clarification in the docs config doc about which resources support dbt_project.yml and yaml settings 
- added an explicit list of properties that _aren't_ supported in the dbt_project.yml page

mirnawong1 committed:
Merge branch 'current' into mwong-macro-yml

mirnawong1 committed:
Merge branch 'current' into mwong-macro-yml

Commit:
tweak snippet and add versionblock

mirnawong1 commented on website/docs/reference/source-properties.md:
```diff
@@ -8,9 +8,13 @@ description: "Learn how to use source properties in dbt."
-Source properties can be declared in `.yml` files in your `models/` directory (as defined by the [`model-paths` config](/reference/project-configs/model-paths)).
 
-You can name these files `whatever_you_want.yml`, and nest them arbitrarily deeply in subfolders within the `models/` directory.
+import PropsCallout from '/snippets/_config-prop-callout.md';
```
i'm pretty sure source properties ARE supported in the `dbt_project.yml` file. but [this doc](https://docs.getdbt.com/reference/configs-and-properties#which-properties-are-not-also-configs) makes it seem like it didn't support it - did i misunderstand the doc?

jtcohen6 replied:
Source **configs** are supported in `dbt_project.yml`. The only relevant source config is `enabled: true|false`. Examples here: https://docs.getdbt.com/reference/source-configs

Source **propertie** are **not** supported in `dbt_project.yml`. They can only be set for specific `sources` / `tables` in properties yaml files

mirnawong1 replied:
oh ok gotcha1

Commit:
remove break

mirnawong1 commented on website/docs/reference/resource-configs/docs.md:
```diff
@@ -70,15 +117,29 @@ snapshots:
 
 <TabItem value="analyses">
 
+You can use the docs property in YAML files, including the `dbt_project.yml`:
```
i'm a little confused why the `dbt_project.yml` is supported for analyses when jerco [said](https://github.com/dbt-labs/docs.getdbt.com/pull/3989#pullrequestreview-1604506567) they can't be configured in the dbt.project.yml file?

jtcohen6 replied:
Correct - you cannot configure multiple `analyses` within `dbt_project.yml`. This isn't for any particularly good reason; it's just never been supported.

They must be configured:
- in-file = `{{ config }}` block in `.sql` file
- in properties yaml file
```yaml
# analyses/properties.yml
analyses:
  - name: my_analysis
    ...
```

mirnawong1 committed:
Merge branch 'current' into mwong-macro-yml

mirnawong1 committed:
Merge branch 'current' into mwong-macro-yml

jtcohen6 commented on website/docs/docs/build/semantic-models.md:
```diff
@@ -8,11 +8,25 @@ sidebar_label: Semantic models
-You can configure semantic models in your dbt project directory in a `YAML` file. Depending on your project structure, you can nest semantic models under a `metrics:` folder or organize them under project sources. 
+<VersionBlock firstVersion="1.7">
+
+You can configure semantic models in your dbt project directory in a `YAML` file, _except_ the `dbt_project.yml` file. Depending on your project structure, you can nest semantic models under a `metrics:` folder or organize them under project sources. 
```
> _except_ the `dbt_project.yml` file

This will change in v1.7:
- https://github.com/dbt-labs/docs.getdbt.com/issues/3997

---

**Update:** I've reread this now, across a few other pages, and I now get your point. You're saying: "You can define properties for sources/models/metrics/etc in a YAML file. You can also _configure_ a number of these resources at once in `dbt_project.yml`. But you cannot actually _define_ their properties within `dbt_project.yml`, even though it is also a YAML file associated with your project." (The same holds for a few other "special" YAML files: `selectors.yml`, `packages.yml`/`dependencies.yml`, `profiles.yml`, ...)

I found this confusing on the first read. I'm trying to think of what else we could do to resolve this ambiguity.

It feels like what we really need is a **standard name** for the kind of yaml file where these things _can_ be defined. I think we've informally settled on something like "properties yaml file":
- https://github.com/dbt-labs/dbt-core/discussions/5104

mirnawong1 replied:
gotcha, thanks so much and apologies for the confusion! you've summarized my understanding totally right and i think you're right about the standard name -- I'll change it to `properties.yml`

mirnawong1 replied:
ok clarified this by updating the partial:

`<span>{props.title} are "special properties" in that you can't configure them in the <code>dbt_project.yml</code> file or using <code>config()</code> blocks. Refer to <a href="https://docs.getdbt.com/reference/configs-and-properties#which-properties-are-not-also-configs" target="_self"> Configs and properties</a> for more info.</span>
`

which turns content into:

![Screenshot 2023-09-18 at 14 24 56](https://github.com/dbt-labs/docs.getdbt.com/assets/89008547/2a2d50d5-969a-42b0-9d39-3f3f655d5fdf)

jtcohen6 commented on website/docs/reference/dbt_project.yml.md:
```diff
@@ -18,12 +18,37 @@ By default, dbt will look for `dbt_project.yml` in your current working director
+- [Exposure properties](/reference/exposure-properties)
+- [Macro properties](/reference/macro-properties)
+- [Analysis properties](/reference/analysis-properties)
+- [Source properties](/reference/source-properties)
```
I think this risks confusing folks. Really, only "configs" can be configured in `dbt_project.yml`. Any "property" that's not a config cannot be configured in `dbt_project.yml`, for any resource type.

In addition:
- `macros` do not support _any_ configs, so they cannot be configured in `dbt_project.yml`
- `analyses` cannot be configured in `dbt_project.yml`
- (v1.6 only) `semantic_models` cannot be configured in `dbt_project.yml` - we've added this capability for v1.7

jtcohen6 commented on website/docs/reference/exposure-properties.md:
```diff
@@ -8,7 +8,11 @@ description: "Read this guide to understand exposure properties in dbt."
+
+import PropsCallout from '/snippets/_config-prop-callout.md';
+
+Exposures are defined in `.yml` files nested under an `exposures:` key. You may define `exposures` in YAML files that also define `sources` or `models`, _except_ in the `dbt_project.yml` file. <PropsCallout title={frontMatter.title}/>  <br /> 
```
> _except_ in the `dbt_project.yml` file.

I found this language confusing - see comment above

jtcohen6 commented on website/docs/reference/resource-configs/docs.md:
```diff
@@ -17,18 +17,35 @@ default_value: {show: true}
+
 <TabItem value="models">
 
+You can use the docs property in YAML files, including the `dbt_project.yml`:
```
It's supported in both places, but with slightly different effects. I might phrase this as:
- You can configure `docs` behavior for many resources at once by setting in `dbt_project.yml`
- You can also use the docs config in properties yaml files, to set/override documentation behaviors for specific resources

jtcohen6 commented on website/docs/reference/resource-configs/docs.md:
````diff
@@ -70,15 +117,29 @@ snapshots:
 
+You can use the docs property in YAML files, including the `dbt_project.yml`:
+
+<File name='dbt_project.yml'>
+
+```yml
+name: "your_dbt_project_name"
+version: "1.0.0"
+config-version: 2
+
+analyses:
+  - name: analysis_name
+    docs:
+      show: true | false
+```
+</File>
````
This code does not work, and should be removed

jtcohen6 commented on website/snippets/_config-prop-callout.md:
```diff
@@ -0,0 +1 @@
+<span>This is because you can set "configs" for many resources in <code>dbt_project.yml</code>, while you can only set "properties" one resource at a time. Since {props.title} are "properties", this means you cannot configure it from <code>dbt_project.yml</code></span>.  Refer to <a href="https://docs.getdbt.com/reference/configs-and-properties#which-properties-are-not-also-configs" target="_self ...
```
Nice snippet!

jtcohen6 reviewed (dismissed):
@mirnawong1 This turned into a bigger lift - thank you for making it. I left a few comments where language was confusing or appeared incorrect.

mirnawong1 committed:
Merge branch 'current' into mwong-macro-yml

mirnawong1 committed:
Merge branch 'current' into mwong-macro-yml

mirnawong1 committed:
Merge branch 'current' into mwong-macro-yml

mirnawong1 committed:
Merge branch 'current' into mwong-macro-yml

mirnawong1 committed:
Merge branch 'current' into mwong-macro-yml

mirnawong1 committed:
Merge branch 'current' into mwong-macro-yml

mirnawong1 committed:
Merge branch 'current' into mwong-macro-yml

Commit:
adding jerco's great feedback

Commit:
clarify

mirnawong1 commented:
hey @jtcohen6 , thank you so much for your guidance and patience here. I've taken some time to understand and made some changes to the language and would love your review. i hope this change clarifies any confusion per your feedback, however, I'm conscious that there still may be some confusion. the following pages are in staging:

- [configs & properties](https://docs-getdbt-com-git-mwong-macro-yml-dbt-labs.vercel.app/reference/configs-and-properties#which-properties-are-not-also-configs)
- [macro properties](https://docs-getdbt-com-git-mwong-macro-yml-dbt-labs.vercel.app/reference/macro-properties)
- [exposure properties](https://docs-getdbt-com-git-mwong-macro-yml-dbt-labs.vercel.app/reference/exposure-properties)
- [analyses properties](https://docs-getdbt-com-git-mwong-macro-yml-dbt-labs.vercel.app/reference/analysis-properties)
- [source properties](https://docs-getdbt-com-git-mwong-macro-yml-dbt-labs.vercel.app/reference/source-properties)
- [docs configs](https://docs-getdbt-com-git-mwong-macro-yml-dbt-labs.vercel.app/reference/resource-configs/docs)

Commit:
change yaml to `properties.yml`

mirnawong1 commented:
picking this up early november

mirnawong1 commented:
picking this back up and reviewing this w/c mon nov 6th

mirnawong1 committed:
Merge branch 'current' into mwong-macro-yml

mirnawong1 commented:
following up this week!

mirnawong1 committed:
Merge branch 'current' into mwong-macro-yml

Commit:
add table

dbeatty10 commented on website/docs/reference/configs-and-properties.md:
```diff
@@ -21,6 +21,21 @@ Whereas you can use **configurations** to:
+| Models        | Describes models and their columns | columns, tests, docs, macros   | tags, materialized, persist_docs, tags, etc. | `properties.yml` -> `.yml file` -> `dbt_project.yml`     |
+| Snapshots     |   Periodic snapshots of data              | tests, macros       | tags, materialized, etc.    |   In-file `config()` block -> `.yml file` -> `dbt_project.yml`            |
+| Seeds         |  Uploaded source data for models         | tests, macros            | tags, persist_docs, etc.     |    In-file `config()` block -> `.yml file` -> `dbt_project.yml`              |
+| Tests         | Asserts "truths" about models    | description, tags, macros       |     In-file `config()` block -> `.yml file` -> `dbt_project.yml`         |
```
Is there content missing from the "Supported properties" or "Supported configurations" columns for the "Tests" row?

If there isn't supposed to be anything in the "Supported configurations" column, then:
```suggestion
| Tests         | Asserts "truths" about models    | description, tags, macros       | |     In-file `config()` block -> `.yml file` -> `dbt_project.yml`         |
```

dbeatty10 commented on website/docs/reference/configs-and-properties.md:
```diff
@@ -21,6 +21,21 @@ Whereas you can use **configurations** to:
 * Declare whether a resource should persist its descriptions as comments in the database
 * Apply tags and "meta" properties
 
+### Resource, properties, and configs table
```
@mirnawong1 How would you feel about splitting this table into a separate PR?

It feels like the rest of the content of this PR is pretty straight-forward and could be reviewed and merged pretty quickly.

This table has a lot more considerations and might take a lot more back-n-forth.

dbeatty10 replied:
Tackling the table separate might also dovetail nicely with https://github.com/dbt-labs/docs.getdbt.com/issues/4314

dbeatty10 replied:
Choosing how to display the inheritance order might also dovetail with https://github.com/dbt-labs/docs.getdbt.com/issues/4371

dbeatty10 replied:
One thing I was thinking about the table:
- might be nice to replace `->` with `→`

Another thing:
- I'd love to have links to each the individual configs and properties referenced within the table.

mirnawong1 replied:
ah ok sounds good @dbeatty10 , i'll do that!

mirnawong1 committed:
Merge branch 'current' into mwong-macro-yml

mirnawong1 committed:
Update configs-and-properties.md

mirnawong1 committed:
Merge branch 'current' into mwong-macro-yml

dbeatty10 committed:
Remove trailing space

dbeatty10 committed:
Merge branch 'current' into mwong-macro-yml

dbeatty10 committed:
Merge branch 'current' into mwong-macro-yml

mirnawong1 committed:
Merge branch 'current' into mwong-macro-yml

dbeatty10 committed:
Merge branch 'current' into mwong-macro-yml

dbeatty10 commented on website/docs/reference/resource-configs/docs.md:
```diff
@@ -45,6 +48,9 @@ This property is not implemented for sources.
 
 <TabItem value="seeds">
 
+You can use the docs property in YAML files, including the `dbt_project.yml`:
+
+
```
What if we remove all the changes related to `website/docs/reference/resource-configs/docs.md` out of this PR and put them into a separate one instead?

We could provide code examples for `dbt_project.yml` similar to [`enabled`](https://docs.getdbt.com/reference/resource-configs/enabled)

Here's some code that should work:

```yaml
name: "my_project"
version: "1.0.0"
config-version: 2
profile: "sandcastle"


models:
  my_project:
    +docs:
      show: false

seeds:
  my_project:
    +docs:
      show: false

snapshots:
  my_project:
    +docs:
      show: false
```

So the generic code examples would look like this:

```yaml
models:
  [<resource-path>](/reference/resource-configs/resource-path):
    +docs:
      show: true | false

seeds:
  [<resource-path>](/reference/resource-configs/resource-path):
    +docs:
      show: true | false

snapshots:
  [<resource-path>](/reference/resource-configs/resource-path):
    +docs:
      show: true | false
```

mirnawong1 replied:
hey @dbeatty10 , i moved it out of this pr and added it here: https://github.com/dbt-labs/docs.getdbt.com/pull/4529

mirnawong1 committed:
Merge branch 'current' into mwong-macro-yml

mirnawong1 commented:
will address monday

mirnawong1 committed:
Merge branch 'current' into mwong-macro-yml

mirnawong1 committed:
Merge branch 'current' into mwong-macro-yml

mirnawong1 committed:
Update docs.md

mirnawong1 committed:
Merge branch 'current' into mwong-macro-yml

dbeatty10 committed:
Merge branch 'current' into mwong-macro-yml

dbeatty10 committed:
Update dbt_project.yml.md

Remove duplicated content

dbeatty10 committed:
Remove extra newline

dbeatty10 reviewed (approved):
Thanks for making these clarifications @mirnawong1  ! 🧠

These previews look good:
- [Configs, properties, what are they?](https://docs-getdbt-com-git-mwong-macro-yml-dbt-labs.vercel.app/reference/configs-and-properties)
- [dbt_project.yml](https://docs-getdbt-com-git-mwong-macro-yml-dbt-labs.vercel.app/reference/dbt_project.yml)

This combined with associated work split out in the following PRs is useful to explain the nuances:
- [#4438](https://github.com/dbt-labs/docs.getdbt.com/pull/4438)
- [#4439](https://github.com/dbt-labs/docs.getdbt.com/pull/4439)
- [#4529](https://github.com/dbt-labs/docs.getdbt.com/pull/4529)

mirnawong1 commented:
woohooo! thank you Doug!

mirnawong1 committed:
Merge branch 'current' into mwong-macro-yml
